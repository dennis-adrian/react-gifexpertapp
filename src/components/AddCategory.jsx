import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('')

    const handleInputChange = (e) => {
        setInputValue(e.target.value);

    }

    const handleSubmit = (e) => {
        //prevenir el comportamiento por defecto del form
        //es decir, prevenir que se recarge
        e.preventDefault();


        if (inputValue.trim().length > 2) {

            //categs representa el valor inicial
            //el valor del arreglo categories
            //que se encuentra en el componoente GifExpertApp
            //y a ese valor le agregamos el input value
            setCategories(categs => [inputValue, ...categs]);
            setInputValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit}>

            <h1>{inputValue}</h1>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
};

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
export default AddCategory;