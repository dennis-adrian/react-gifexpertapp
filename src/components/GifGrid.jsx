import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const { data: images, loading } = useFetchGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>

            {/* esto es como un operador ternario
                la ventaja es que solo valida por true
                no es necesario un false */}
            {loading && <p>Loading...</p>}

            <div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                    ))
                }
            </div>
        </>
    )
}
