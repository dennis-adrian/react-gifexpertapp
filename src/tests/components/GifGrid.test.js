import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
//de esta manera finjimos la llamada a este archivo
jest.mock('../../hooks/useFetchGifs')

describe('Testing on <GifGrid />', () => {

    const category = 'Dragon Ball'

    test('should be shown correctly', () => {

        //finjimos que estos son los datos que está recolectando useFetchGifs
        //desde la api (ya que en realidad, no estamos llamando a la api, hay 
        //que finjirlo)
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    })

    test('should show items when images are loaded with useFetchGifs', () => {

        //el mock sirve  para que el componente crea que tiene información
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/algo',
            title: 'Cualquier cosa'
        },
        {
            id: 'ABC',
            url: 'https://localhost/algo',
            title: 'Cualquier cosa'
        },
        {
            id: 'ABC',
            url: 'https://localhost/algo',
            title: 'Cualquier cosa'
        }]

        //finjimos que estos son los datos que está recolectando useFetchGifs
        //desde la api (ya que en realidad, no estamos llamando a la api, hay 
        //que finjirlo)
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category} />);

        expect(wrapper).toMatchSnapshot();
        //evaluamos que el parrafo de "loading" no exista
        //porque cuando los datos ya están cargados, ese párrafo no debería existir
        expect(wrapper.find('p').exists()).toBe(false);
        //evaluamos que la cantidad de items coincida con la cantidad de objetos en el array
        //que simula ser el fetch a la api
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);
    })

})

