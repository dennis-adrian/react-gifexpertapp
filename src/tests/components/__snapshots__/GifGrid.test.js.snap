// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <GifGrid /> should be shown correctly 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Dragon Ball
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Testing on <GifGrid /> should show items when images are loaded with useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Dragon Ball
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/algo"
    />
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/algo"
    />
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="https://localhost/algo"
    />
  </div>
</Fragment>
`;
